src/Native/Unix/System.Native/pal_io.c

```diff
@@ -324,41 +324,38 @@ static void ConvertDirent(const struct dirent* entry, struct DirectoryEntry* out
 #endif
 }
 
-int32_t SystemNative_GetDirentSize(void)
+int32_t SystemNative_GetReadDirRBufferSize(void)
 {
+#if HAVE_READDIR_R
     // dirent should be under 2k in size
     assert(sizeof(struct dirent) < 2048);
     return sizeof(struct dirent);
+#else
+    return 0;
+#endif
 }
 
-// To reduce the number of string copies, this function calling pattern works as follows:
-// 1) The managed code calls GetDirentSize() to get the platform-specific
-//    size of the dirent struct.
-// 2) The managed code creates a byte[] buffer of the size of the native dirent
-//    and passes a pointer to this buffer to this function.
-// 3) This function passes input byte[] buffer to the OS to fill with dirent
-//    data which makes the 1st strcpy.
-// 4) The ConvertDirent function will fill DirectoryEntry outputEntry with
-//    pointers from byte[] buffer.
-// 5) The managed code uses DirectoryEntry outputEntry to find start of d_name
-//    and the value of d_namelen, if avalable, to copy the name from
-//    byte[] buffer into a managed string that the caller can use; this makes
-//    the 2nd and final strcpy.
+// To reduce the number of string copies, the caller of this function is responsible to ensure the memory
+// referenced by outputEntry remains valid until it is read.
+// If the platform supports readdir_r, the caller provides a buffer into which the data is read.
+// If the platform uses readdir, the caller must ensure no calls are made to readdir/closedir since those will invalidate
+// the current dirent.
```

Comment:
We also rely on readdir being thread-safe.  If it's not, bad things happen.

Reply:
We were relying on that already. The glibc readdir_r man page is wording this _vaguely_ as follows:

> In the current POSIX.1 specification (POSIX.1-2008), readdir(3) is not required to be thread-safe.  However, in modern implementations (including the glibc implementation),  concurrent  calls  to
          readdir(3)  that  specify different directory streams are thread-safe.  Therefore, the use of readdir_r() is generally unnecessary in multithreaded programs.  In cases where multiple threads must
          read from the same directory stream, using readdir(3) with external synchronization is still preferable to the use of readdir_r(), for the reasons given in the points above.

I think it is safe to assume readdir is thread-safe on platforms where readdir_r is not available, as long as the user does not concurrently call readdir on the same DIR.

Reply:
> We were relying on that already.

I know, I was suggesting it should be added to the comment :)